allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -250,18 +251,24 @@ def __init__(self,
         super(WikiTablesDecoderStep, self).__init__()
         # Decoder output dim needs to be the same as the encoder output dim since we initialize the
         # hidden state of the decoder with that of the final hidden states of the encoder.
-        self._output_dim = encoder_output_dim
+        self.vocab = vocab
         self._action_embedder = Embedding(num_actions, action_embedding_dim)
         self._input_attention = Attention(attention_function)
         self._start_index = start_index
 
-        # The output of attention, which is a weighted average over encoder outputs, will be
-        # concatenated to the input vector of the decoder at each time step.
-        self._decoder_input_dim = encoder_output_dim + action_embedding_dim
+        output_dim = encoder_output_dim
+        input_dim = output_dim
+        # Our decoder input will be the concatenation of the decoder hidden state and the previous
+        # action embedding, and we'll project that down to the decoder's `input_dim`, which we
+        # arbitrarily set to be the same as `output_dim`.
+        self._input_projection_layer = Linear(output_dim + action_embedding_dim, input_dim)
+        # Before making a prediction, we'll compute an attention over the input given our updated
+        # hidden state.  Then we concatenate that with the decoder state and project to
+        # `num_actions` to make a prediction.
```

Comment:
Why concatenate the attention over input before projection? That is not a common practice, is it?

Reply:
We had several long conversations about this in other places, but the relevant outcome of those conversations is that these equations are slightly different from what most papers I've seen have done, but they are now closer to current best practices than what we had before, and the difference between what's here and what's in other papers is this has an extra use of the attention from the previous timestep.